From ef2823cbf1019a4984312fabadde13b54d2af0f1 Mon Sep 17 00:00:00 2001
From: Sarika Sharma <quic_sarishar@quicinc.com>
Date: Thu, 15 Feb 2024 11:30:37 +0530
Subject: [PATCH] wifi: ath12k: Add memset and update default rate value in wmi
 tx completion

When AP/STA + Monitor interfaces enabled together, ieee80211_tx_status()
is getting called from tx completion handler for data frames and from
wmi tx completion handler for management frames. Mac80211 copies these
skbs to the monitor interface after updating the statistics. Further
mac80211 uses skb->cb for generating radio tap header and it is having
junk values causing invalid radiotap length. This issue observed only in
WMI tx completion path whereas memset happens properly in data path.

Fix it by adding memset and update default rate values before calling
ieee80211_tx_status().

Signed-off-by: P Praneesh <quic_ppranees@quicinc.com>
Signed-off-by: Sarika Sharma <quic_sarishar@quicinc.com>
---
 drivers/net/wireless/ath/ath12k/wmi.c | 8 ++++++++
 1 file changed, 8 insertions(+)

diff --git a/drivers/net/wireless/ath/ath12k/wmi.c b/drivers/net/wireless/ath/ath12k/wmi.c
index e8b2de2..75d8f32 100644
--- a/drivers/net/wireless/ath/ath12k/wmi.c
+++ b/drivers/net/wireless/ath/ath12k/wmi.c
@@ -7865,6 +7865,10 @@ skip_mgmt_stats:
 	spin_unlock_bh(&ar->data_lock);
 
 	info = IEEE80211_SKB_CB(msdu);
+	memset(&info->status, 0, sizeof(info->status));
+
+	info->status.rates[0].idx = -1;
+
 	if ((!(info->flags & IEEE80211_TX_CTL_NO_ACK)) && !status) {
 		info->flags |= IEEE80211_TX_STAT_ACK;
 		info->status.ack_signal = ack_rssi;
@@ -7945,6 +7949,10 @@ static void wmi_process_offchan_tx_comp(struct ath12k *ar, u32 desc_id,
 	spin_unlock_bh(&ar->data_lock);
 
 	info = IEEE80211_SKB_CB(msdu);
+	memset(&info->status, 0, sizeof(info->status));
+
+	info->status.rates[0].idx = -1;
+
 	if (!(info->flags & IEEE80211_TX_CTL_NO_ACK) && !status)
 		info->flags |= IEEE80211_TX_STAT_ACK;
 
-- 
2.34.1

